### PR TITLE
Gradient checkpointing BERT & ALBERT poc

### PR DESCRIPTION
**Pull request by LysandreJik**
Proof of concept for gradient checkpointing in PyTorch, using a model-agnostic approach. The POC is done for BERT and ALBERT.

Pros:

- Model agnostic, only a few lines to add to models to be able to use this functionality
- Reinforces the model layer API, adding `get_layers()` (name to be discussed) alongside `get_input_embeddings()` and `get_output_embeddings()`

Cons:

- The checkpoint API can only handle positional arguments, pytorch tensors or None only. This means that:
    - The `output_hidden_states` must be cast to a tensor in the model
    - Models that pass keyword arguments to their layers need to pass positional arguments (see GPT-2 for example, which uses keyword arguments [here](https://github.com/huggingface/transformers/blob/b45e65efa0fbff2611ddd68e14fa75cacef3fe08/src/transformers/modeling_gpt2.py#L488-L493)).

If you think this is a cool API, I'll go ahead and implement this for the remaining models. @patrickvonplaten @thomwolf @julien-c @sgugger @ibeltagy

Here are the results using the benchmarking script:

```py
from transformers import PyTorchBenchmark, PyTorchBenchmarkArguments, BertConfig
args = PyTorchBenchmarkArguments(models=["bert-base-cased"], batch_sizes=[8], sequence_lengths=[8, 32, 128, 512], no_inference=True, training=True)
config_base = BertConfig.from_pretrained("bert-base-cased", gradient_checkpointing=False)
benchmark = PyTorchBenchmark(args, configs=[config_base])
benchmark.run()
```
Result (only relevant info):

```
====================        TRAIN - SPEED - RESULTS         ====================
--------------------------------------------------------------------------------
          Model Name             Batch Size     Seq Length     Time in s   
--------------------------------------------------------------------------------
       bert-base-cased               8               8             0.028     
       bert-base-cased               8               32            0.029     
       bert-base-cased               8              128            0.072     
       bert-base-cased               8              512            0.296     
--------------------------------------------------------------------------------

====================        TRAIN - MEMORY - RESULTS        ====================
--------------------------------------------------------------------------------
          Model Name             Batch Size     Seq Length    Memory in MB 
--------------------------------------------------------------------------------
       bert-base-cased               8               8              2419     
       bert-base-cased               8               32             2481     
       bert-base-cased               8              128             2985     
       bert-base-cased               8              512             8233     
--------------------------------------------------------------------------------
```

```py
from transformers import PyTorchBenchmark, PyTorchBenchmarkArguments, BertConfig
args = PyTorchBenchmarkArguments(models=["bert-base-cased"], batch_sizes=[8], sequence_lengths=[8, 32, 128, 512], no_inference=True, training=True)
config_base = BertConfig.from_pretrained("bert-base-cased", gradient_checkpointing=True)
benchmark = PyTorchBenchmark(args, configs=[config_base])
benchmark.run()
```

Result (only relevant info):

```
====================        TRAIN - SPEED - RESULTS         ====================
--------------------------------------------------------------------------------
          Model Name             Batch Size     Seq Length     Time in s   
--------------------------------------------------------------------------------
       bert-base-cased               8               8             0.049     
       bert-base-cased               8               32             0.05     
       bert-base-cased               8              128            0.109     
       bert-base-cased               8              512            0.473     
--------------------------------------------------------------------------------

====================        TRAIN - MEMORY - RESULTS        ====================
--------------------------------------------------------------------------------
          Model Name             Batch Size     Seq Length    Memory in MB 
--------------------------------------------------------------------------------
       bert-base-cased               8               8              2385     
       bert-base-cased               8               32             2403     
       bert-base-cased               8              128             2465     
       bert-base-cased               8              512             3969     
--------------------------------------------------------------------------------
```

### CODE REVIEW

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -676,6 +677,31 @@ def test_inputs_embeds(self):
...
+            model_with_gc.load_state_dict(model.state_dict())
+            model_with_gc.to(torch_device)
+            model_with_gc.eval()
+            outputs_with_checkpointing = model_with_gc(**self._prepare_for_class(inputs_dict, model_class))
+
+            for output_no_checkpointing, output_with_checkpointing in zip(
+                outputs_no_checkpointing, outputs_with_checkpointing
+            ):
+                if isinstance(output_with_checkpointing, torch.Tensor):
+                    self.assertTrue(torch.allclose(output_no_checkpointing, output_with_checkpointing))
```
Comment (LysandreJik) [inline tests/test_modeling_common.py:703]: A better test would probably be to additionally test that the memory footprint is lower.

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -456,6 +475,11 @@ def init_weights(self):
         # Tie weights if needed
         self.tie_weights()
 
+        # Gradient accumulation if needed
+        if self.config.gradient_checkpointing:
+            for layer in self.get_layers():
+                layer.register_forward_pre_hook(self._gradient_accumulation_hook)
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:481]: Why not just replace the forward method here? I don't see what the hook adds since it seems like the only thing it does is:
```
layer.forward = functools.partial(torch.utils.checkpoint.checkpoint, layer.forward)
```

**Review by sgugger — APPROVED** [review]
It's nice and easy though I don't understand why we need a hook.

File: src/transformers/modeling_utils.py (hunk 2 above)
Comment (LysandreJik) [reply src/transformers/modeling_utils.py]: You're correct, I was mistaken here.

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -345,6 +346,20 @@ def get_output_embeddings(self):
...
 
+    def get_layers(self):
+        """
+        Returns the model's transformer layers.
+
+        Returns:
+            :obj:`List` or :obj:`nn.ModuleList`:
+                A list or :obj:`nn.ModuleList` containing all transformer layers.
+        """
+        base_model = getattr(self, self.base_model_prefix, self)
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:357]: Really hope that this works for all models! The only model where I see that it doesn't work at the moment is T5, but maybe we can just overwrite it there?

**Comment by patrickvonplaten** [comment]
I really like the API, I think it's fine if we enforce all attention layers to use positional arguments and wrap the output attentions bool into a tensor. 

Can we test how much memory is saved here for `bert-base-uncased` layers 6 - 18 for example? 
Should be quite easy to do now with the benchmark utils.

**Comment by LysandreJik** [comment]
Benchmarked with the script and updated the PR @patrickvonplaten!

**Comment by patrickvonplaten** [comment]
Awesome, it looks we can gain quite a lot of memory :-)

**Review by patrickvonplaten — APPROVED** [review]

File: src/transformers/modeling_bert.py [hunk 4]
```diff
@@ -409,35 +409,19 @@ def forward(
     ):
         all_hidden_states = ()
         all_attentions = ()
+        output_attentions = torch.tensor(output_attentions, dtype=torch.bool)
```
Comment (ibeltagy) [inline src/transformers/modeling_bert.py:412]: Is it correct to have this tensor on the cpu or is `torch.utils.checkpoint.checkpoint` expecting everything to be on the same device? Also, it might be worth trying it with ddp just to be sure. 

Another thing to check is pytorch-xla. It probably doesn't support gradient checkpointing in the first place, but just checking that a mix of cpu/tpu tensors is not an issue.

Comment (ibeltagy) [reply src/transformers/modeling_bert.py:412]: Alternatively, is it possible to generalize the following wrapper of the forward function to automatically handle non-tensor arguments? 
```
def create_custom_forward(module):
    def custom_forward(*inputs):
        return module(*inputs, output_attentions)
    return custom_forward
```

File: src/transformers/modeling_albert.py [hunk 5]
```diff
@@ -327,6 +327,7 @@ def forward(
         hidden_states = self.embedding_hidden_mapping_in(hidden_states)
 
         all_attentions = ()
+        output_attentions = torch.tensor(output_attentions, dtype=torch.bool)
```
Comment (LysandreJik) [inline src/transformers/modeling_albert.py:330]: This is not compatible with torch <= 1.2. Will find a way to make it compatible.

Comment (patrickvonplaten) [reply src/transformers/modeling_albert.py:330]: Is our lib still compatible with torch <= 1.2 ? I feel like there are many cases where torch 1.2 does not work anymore

**Review by ibeltagy — COMMENTED** [review]
This is great, thanks @LysandreJik. Two things I want to mention, 

- An alternative solution to casting non-tensor arguments is to have another wrapper around the forward function that decides which arguments to pass to `torch.utils.checkpoint.checkpoint`. We have an example in the BERT code but I am not sure how to generalize it to other models. 

- Regarding this issue, https://github.com/huggingface/transformers/pull/4659#issuecomment-643858093, for ddp to work with gradient checkpointing, all the model parameters should contribute to the computation of the loss. We probably need a general solution that finds unused model parameters and removes them from the model.

**Review by thomwolf — APPROVED** [review]
I like this!

**Comment by ibeltagy** [comment]
@LysandreJik , another problem is that `torch.utils.checkpoint.checkpoint` expects the function to return one object of type `Variable`. This won't work with forward functions that return more than one object as in [here](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_bart.py#L321).

**Comment by ibeltagy** [comment]
@LysandreJik , any plans to resurrect this?

**Comment by LysandreJik** [comment]
Yes it's on my TODO (probably in ~2 weeks), and will be for most models (with some exceptions like BART and T5, which need a lot of plumbing to work with this POC)

File: src/transformers/modeling_albert.py (hunk 5 above)
Comment (sgugger) [reply src/transformers/modeling_albert.py:330]: If cast as a tensor of ints, is it breaking many things in the PR?

File: src/transformers/modeling_albert.py [hunk 6]
```diff
@@ -462,6 +463,9 @@ def get_input_embeddings(self):
     def set_input_embeddings(self, value):
         self.embeddings.word_embeddings = value
 
+    def get_layers(self):
```
Comment (sgugger) [inline src/transformers/modeling_albert.py:466]: I like this new entry point and feel it could be useful for other new additions to the lib!

File: src/transformers/modeling_utils.py [hunk 7]
```diff
@@ -345,6 +346,20 @@ def get_output_embeddings(self):
         """
         return None  # Overwrite for models with output embeddings
 
+    def get_layers(self):
+        """
+        Returns the model's transformer layers.
+
+        Returns:
+            :obj:`List` or :obj:`nn.ModuleList`:
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:354]:
```suggestion
            :obj:`List[nn.Module]` or :obj:`nn.ModuleList`:
```

File: src/transformers/modeling_utils.py (hunk 3 above)
Comment (sgugger) [reply src/transformers/modeling_utils.py:357]: I think overwriting for the few models that need it is fine.

**Review by sgugger — APPROVED** [review]
Looks very cool, thanks for all the work on this!
